### DISCUSSION TITLE
[v5] Background job handling solutions

### DISCUSSION THREAD

**Discussion by watzon** [adonisjs/core#1992] [Help]
I'm attempting (and maybe this is a bad idea) to write a production application for the company I work for using Adonis v5. The original plan was to use Ruby on Rails, but my team is more experienced with JS and I've been aching for a chance to use Adonis is a production app.

This particular application is for tracking KPIs, the data from which comes from our store backend. Basically there will be queries which run at specific intervals and add the gathered metrics to a database to use for data visualization and marketing purposes. This, of course, means I need some way to create jobs which run at specific intervals so that queries can run when they're supposed to.

So I'm wondering if anyone else has found themselves in need of something similar and what solutions you decided to implement. It would be nice to have something integrated, like Rails' Active Job, but obviously we don't have that yet.

**Comment by thetutlage** [comment]
If you are confused between v4 and v5, then I will suggest 100% go with v5. If its between Rails or AdonisJS, then you are best one to make the decisions, since I am the creator of the framework and answer will be biased.

Yes, there is no out of the box support for Queues in AdonisJS v5, but you can use [bullmq](https://docs.bullmq.io), one of the most popular option for using queues in Node.js land.

**Comment by carsonwright** [comment]
I would recommend beequeue personally. I've used bull and it has some weird
stuff. We switched to beequeue from bull a couple years ago.

**Reply by thetutlage** [reply]
Do you remember the exact issues you faced? I have personally used bee queue and it is a wonderful package. However, it is slim on the number of features in comparison to bull and for the framework, I have to opt for something feature rich. 

So it will be great to know the issues others have faced with bull, so that I can make informed decisions when opting for it

**Comment by carsonwright** [comment]
We had issues with double runs actually. A job happening twice every now
and then. It was super rare but our system runs hundreds of jobs a second
so we started to see it. It was also slower. We built our own ui for
beequeue to track and everything. It's totally what you want though. It's
been two years so I'm a little fuzzy on it.

**Comment by jpedroschmitz** [comment]
Hey, I'm a maintainer from [@rocketseat/adonis-bull](https://github.com/Rocketseat/adonis-bull) and we have released the next version of the library, which is ready for Adonis v5. We have been working hard to make it super easy to set up and have out of the box support for Queues (under the hood we use `bullmq`) 💜 

To get started with it, you can install it with NPM or Yarn:

```sh
yarn add @rocketseat/adonis-bull@alpha
# or
npm install @rocketseat/adonis-bull@alpha
```

After installing it, run `node ace invoke @rocketseat/adonis-bull`, choose a start option and you're ready to go ⚡️

You can find the complete documentation of the package [here](https://github.com/Rocketseat/adonis-bull/tree/alpha#readme).
